espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/SourceDocument.kt

```diff
@@ -0,0 +1,316 @@
+                    serializeView(rootView, 0)
+                    serializer!!.endDocument()
+                    logger.info(String.format("The source XML tree has been fetched in %sms using %s",
+                            SystemClock.uptimeMillis() - startTime, streamType.simpleName))
+                } catch (e: OutOfMemoryError) {
+                    lastError = e
+                    continue
+                } finally {
+                    outputStream.close()
+                }
+                return if (outputStream is FileOutputStream)
+                    getApplicationContext<Context>().openFileInput(tmpXmlName)
+                else
+                    ByteArrayInputStream((outputStream as ByteArrayOutputStream).toByteArray())
+            } catch (e: IOException) {
+                lastError = e
+            }
+
+        }
+        if (lastError is OutOfMemoryError) {
+            throw lastError
+        }
+        throw AppiumException(lastError)
+    }
+
+    private fun performCleanup() {
+        if (tmpXmlName != null) {
+            getApplicationContext<Context>().deleteFile(tmpXmlName)
+            tmpXmlName = null
+        }
+    }
+
+    @Throws(AppiumException::class)
+    fun toXMLString(): String {
+        try {
+            RESOURCES_GUARD.acquire()
+        } catch (e: InterruptedException) {
+            throw AppiumException(e)
+        }
+
+        try {
+            toStream().use { xmlStream ->
+                val sb = StringBuilder()
+                val reader = BufferedReader(InputStreamReader(xmlStream, XML_ENCODING))
+                var line: String? = reader.readLine()
+                while (line != null) {
+                    sb.append(line)
+                    line = reader.readLine()
+                }
+                return sb.toString()
+            }
+        } catch (e: IOException) {
+            throw AppiumException(e)
+        } finally {
+            performCleanup()
+            RESOURCES_GUARD.release()
+        }
+    }
+
+    @Throws(AppiumException::class)
+    fun findViewsByXPath(xpathSelector: String): List<View> {
+        var list: NodeList? = null
+        try {
+            // Get the Nodes that match the provided xpath
+            val expr = xpath.compile(xpathSelector)
+            try {
+                RESOURCES_GUARD.acquire()
+            } catch (e: InterruptedException) {
+                throw AppiumException(e)
+            }
+
+            try {
+                toStream().use { xmlStream -> list = expr.evaluate(InputSource(xmlStream), XPathConstants.NODESET) as NodeList }
+            } catch (e: IOException) {
+                throw AppiumException(e)
+            } finally {
+                performCleanup()
+                RESOURCES_GUARD.release()
+            }
+
+            // Get a list of elements that are associated with that node
+            val views = ArrayList<View>()
+            if (list != null) {
```

Comment:
let

Reply:
actually, this assignment can be done within one single reduce, like
`views = list?.reduce(...)`